lib/sqlalchemy/sql/functions.py

```diff
@@ -49,7 +55,48 @@ def register_function(identifier, fn, package="_default"):
 
     """
     reg = _registry[package]
-    reg[identifier] = fn
+    case_sensitive_reg = _case_sensitive_registry[package]
+    raw_identifier = identifier
+    identifier = identifier.lower()
+
+    # Check if a function with the same lowercase identifier is registered.
+    if identifier in reg:
+        if raw_identifier in case_sensitive_reg[identifier]:
+            warnings.warn(
+                "The GenericFunction '{}' is already registered and "
+                "is going to be overriden.".format(identifier),
+                sa_exc.SAWarning)
+            reg[identifier] = fn
+        else:
+            # If a function with the same lowercase identifier is registered,
+            # then these 2 functions are considered as case-sensitive.
+            # Note: This case should raise an error in a later release.
+            reg.pop(identifier)
```

Comment:
**mike bayer** (zzzeek) wrote:

reg.pop(identifier, None) in case two threads are here after both passing the check.

[2] However, I have an idea which is to add a little more information.   We instead say:

   reg[identifier] = _CASE_SENSITIVE

Where _CASE_SENSITIVE is a util.symbol() defined at the module level.   This gives us a clue to make a quicker decision during the lookup and also is less ambiguous.

Reply:
I like this design, it is much more clear and explicit. I update it.